File: Problems/GrayCode.py
def grayCode(n):
    if n == 0:
        return [0]
    nMax = 2**n
    return [(x >> 1) ^ x for x in range(nMax)]


print(grayCode(4))

# (0 >> 1) ^ 0 = 000
# (1 >> 1) ^ 1 = 001
# (2 >> 1) ^ 2 = 01 ^ 10 = 011
# (3 >> 1) ^ 3 = 01 ^ 11 = 010
# (4 >> 1) ^ 4 = 010 ^ 100 = 110
# (5 >> 1) ^ 5 = 010 ^ 101 = 111

File: Problems/GrayCode.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <bitset>
using namespace std;

class Solution
{
    // iterative solution
public:
    vector<int> grayCode(int n)
    {
        unordered_set<int> uset;
        vector<int> res;
        res.push_back(0);
        uset.insert(0);
        while (res.size() != (1 << n))
        {
            int prev = res.at(res.size() - 1);
            for (int i = 0; i < n; i++)
            {
                int candidate = prev ^ (1 << i);
                if (uset.find(candidate) == uset.end())
                {
                    res.push_back(candidate);
                    uset.insert(candidate);
                    break;
                }
            }
        }
        return res;
    }

    // recursive backtracking solution using bitset
    void utils(bitset<32> &bits, vector<int> &result, int k)
    {
        if (k == 0)
        {
            result.push_back(bits.to_ulong());
        }
        else
        {
            utils(bits, result, k - 1);
            bits.flip(k - 1);
            utils(bits, result, k - 1);
        }
    }

public:
    vector<int> grayCode(int n)
    {
        bitset<32> bits;
        vector<int> result;
        utils(bits, result, n);
        return result;
    }
};